### DISCUSSION TITLE
How can I put my cross-compiled libffi somewhere micropython unix port will find it?

### DISCUSSION THREAD

**Discussion by JiveyGuy** [micropython/micropython#13012] [Libraries & Drivers] [answered]
Related to: 

Hello, I am trying to cross-compile to an arm-linux-gnueabihf host from x86_64-linux PC. 

I used these steps: 

```bash
cd micropython/mpy-cross
make
cd ../ports/unix
make submodules
make CROSS_COMPILE=arm-linux-gnueabihf-
```

Which generates this error: 
```bash
modffi.c:40:10: fatal error: ffi.h: No such file or directory
 #include <ffi.h>
```

I found this discussion on the old forum archive related to this error: 
[Discussion](https://forum.micropython.org/viewtopic.php?t=1648)

I followed the work around of providing `MICROPY_PY_FFI=0` compile flag which **worked**. 

However, this can lead to issues because now libraries that are called through libffi do not work. So following the advice of the forum user dhylands:
![Screenshot from 2023-11-17 18-09-42](https://github.com/micropython/micropython/assets/20270893/b59d0a67-32cc-4cfc-b1a8-2ee3ede9af7a)

I cross-compiled libffi with these steps:
```bash
git clone https://github.com/libffi/libffi.git
cd libffi
./autogen.sh
./configure CC=arm-linux-gnueabihf-gcc CXX=arm-linux-gnueabihf-g++ -host=arm-linux -build=x86_64-linux -prefix=/home/jivey/Devel/state_machine/install_files -disable-dtrace -disable-systemtap -disable-selinux
sudo cp arm-unknown-linux-gnu/include/ffi.h /usr/include
sudo cp arm-unknown-linux-gnu/include/ffitarget.h /usr/include
sudo cp arm-unknown-linux-gnu/fficonfig.h /usr/include
cp -r arm-unknown-linux-gnu/ ../micropython/lib/libffi
```

However, this did not work because make could not find libffi.h still after these steps. What am I doing wrong or where should I copy these files so make will see them while cross-compiling?

**Comment by jimmo** [comment, marked as answer]
@JiveyGuy 

One option is to do `make MICROPY_STANDALONE=1` which will use a locally built copy of libffi from a submodule and set the appropriate compiler flags to use that. (You may need to update the submodule to a more recent version if you run into issues).

Otherwise look at ports/unix/Makefile and you might be able to adapt the handling of MICROPY_STANDALONE to work with your locally built copy.

**Reply by JiveyGuy** [reply]
Awesome reply! Was about to swap to C before this, for people encountering this issue I used these steps: 

===================================



Need to get a cross compiler for ArmV7, I installed: 

* gcc-linaro-7.5.0-2019.12-x86_64_arm-linux-gnueabihf

First follow: https://kclin.tw/2013/12/17/compile-libffi-for-embedded-arm/

Setps:
```bash
git clone https://github.com/libffi/libffi.git
cd libffi
./autogen.sh
./configure CC=arm-linux-gnueabihf-gcc CXX=arm-linux-gnueabihf-g++ -host=arm-linux -build=x86_64-linux -prefix=/home/jivey/Devel/state_machine/install_files -disable-dtrace -disable-systemtap -disable-selinux
cd arm-linux-gnueabihf-gnu
make 
make install
```


Following: https://community.element14.com/members-area/personalblogs/b/andy-clark-s-blog/posts/cross-compiling-micropython-on-wsl-for-stm32

steps:
```bash
git clone https://github.com/openmv/micropython.git
cd micropython/mpy-cross
make
cd ../ports/unix
make submodules
mkdir build-standard/lib/libffi/out/
cp -r ../../lib build-standard/lib/libffi/out/
make CROSS_COMPILE=arm-linux-gnueabihf- V=1 unix_ffi=True MICROPY_STANDALONE=1
```
